File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.py
#解题思路
#平衡二叉搜索树的要求是每个节点左右子树的高度差最多为1。那只要取序列的中间数作为根节点，
#左边的序列再组成它的左子树，右边的序列组成它的右子树，递归完成构造。

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        return self._sortedArrayToBST(nums, 0, len(nums))

    def _sortedArrayToBST(self, nums, left, right):
        if left == right:
            return None
        mid = (left + right) >> 1
        root = TreeNode(nums[mid])
        root.left = self._sortedArrayToBST(nums, left, mid)
        root.right = self._sortedArrayToBST(nums, mid + 1, right)
        return root


if __name__ == "__main__":
    None

File: 108. Convert Sorted Array to Binary Search Tree/108. Convert Sorted Array to Binary Search Tree.java
/*
本题解题思路还是从递归出发，但是要结合二叉查找树的定义，每次将数组二分中间值作为根节点，
依次递归二分并将中间值赋值给根节点，最后返回根结点，递归返回的临界条件是：左索引大于右索引
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        if(nums.size()==0)
            return NULL;
        return sortedSub(nums,0,nums.size()-1);
        
    }
private: 
    TreeNode* sortedSub(vector<int>& nums, int l, int r)  {  
    if(l>r)  
        return NULL;  
    int m = (l+r)/2;  
    TreeNode* root = new TreeNode(nums[m]);  
    root->left = sortedSub(nums,l,m-1);  
    root->right = sortedSub(nums,m+1,r);  
    return root;  
}  
};
